### PR TITLE
51 get examples of custom rules

### LINKED ISSUE #51: Extend DSL for user-defined rules
Example
image

- [ ] Support user functions like `concat`
  - [ ] get examples from @yegor256, @maxonfjvipon
  - [ ] choose a special prefix
- [ ] Support correct attributes renaming
  - rename all occurences
  - fresh names
  - "escape analysis"
- [ ] Support optional unwrapping of a rule result
  - inline RULE definition

### PR DESCRIPTION
**Pull request by deemp**
Closes #51 

Links:
- [original BNFC grammar](https://github.com/objectionary/normalizer/blob/51-collect-examples-of-user-defined-functions/eo-phi-normalizer/grammar/EO/Phi/Syntax.cf)
- [use cases](https://github.com/objectionary/normalizer/blob/51-collect-examples-of-user-defined-functions/user-defined-functions/use-cases.yaml)
- [specifications](https://github.com/objectionary/normalizer/blob/51-collect-examples-of-user-defined-functions/user-defined-functions/specifications.yaml)

---

## PR-Codex overview
This PR enhances the specifications.yaml file by defining MetaBinding patterns and functions for manipulating bindings.

### Detailed summary
- Defined MetaBinding patterns for matching bindings
- Added functions for manipulating bindings like $replace and $combine

> The following files were skipped due to too many changes: `user-defined-functions/use-cases.yaml`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: user-defined-functions/specifications.yaml [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      - ⟦ a ↦ ?B ⟧
+  - pattern: |
+      ExpressionFunction. Expression ::= Function
+    examples:
+      - $name(?a)
+functions:
+  - pattern: |
+      FunctionMt. Function ::= "$mt" "(" <prefix :: (Expression)> "," <suffix :: (Expression)> ")"
+    description: |
+      Return the string "<prefix>.<suffix>".
```
Comment (fizruk) [inline user-defined-functions/specifications.yaml:58]: I do not think we should work with strings. We should work with AST. `<prefix>` should be a valid $\varphi$-term and `<suffix>` should be a valid attribute (or chain of dispatch/application).

Also naming is not explained. Why `mt`?

Comment (fizruk) [reply user-defined-functions/specifications.yaml:58]: And also, why do we need `$mt` if we can simply use dot `.`?

File: user-defined-functions/specifications.yaml [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    examples:
+      - applications:
+          - application: $mt(ξ, $mt(ρ, $mt($name(?b_a), write(α0 ↦ ?t))))
+            # Note the result isn't a value because the result contains MetaId-s
+            result: ξ.ρ.?b_a.write(α0 ↦ ?t)
+  - pattern: |
+      FunctionName. Function ::= "$name" "(" <binding :: (MetaId)> ")"
+    description: |
+      Return the name of <binding>.
```
Comment (fizruk) [inline user-defined-functions/specifications.yaml:67]: Why do we need it if we have `?a` for the name already?

File: user-defined-functions/specifications.yaml (hunk 1 above)
Comment (deemp) [reply user-defined-functions/specifications.yaml:58]: > I do not think we should work with strings.

Yes, it should be a dispatch.

> Also naming is not explained. Why mt?

IDK. It comes from use cases. I'd prefer $dot.

>  And also, why do we need $mt if we can simply use dot?

I believe we shouldn't use dot with functions, should we?

1. `a.$name(b)`
1. `$dot(a, $name(b))`

File: user-defined-functions/specifications.yaml (hunk 2 above)
Comment (deemp) [reply user-defined-functions/specifications.yaml:67]: We may not have the name in case of `\?A ?a..b -> c`

Comment (deemp) [reply user-defined-functions/specifications.yaml:67]: Also, we can use $name to extract all names from a list of bindings.
We may need to do that when we need to check that a name of some binding doesn't belong to that list.

**Comment by deemp** [comment]
Closing because this PR was split into https://github.com/objectionary/normalizer/pull/259 and https://github.com/objectionary/normalizer/pull/258.

**Pull request closed without merging**